Fixes for RZ formulation for INS/INSAD

lindsayad opened the pull request:
Closes #25307

lindsayad committed:
Fix hard-coding of radial coordinates in INS/INSAD objects

Also:
- Remove deprecated Problem/RZ syntax
- Add hand-coded Jacobian Dirichlet x-radial-coord test
- Add hand-coded Jacobian natural outlet bc test
- Add unstabilized tests

Closes #25307

lindsayad committed:
Add ability to control whether we identify variable groups

lindsayad committed:
Add XYZ LBB-stable w/ SUPG and traction form tests

lindsayad committed:
Now fixing the traction form

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1884546/) on 72293b4 : invalidated by @lindsayad

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1884539/) on 72293b4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26072/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1884556/) on 72293b4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>26072a</th>
      <th colspan=4>#26072 72293b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.74%</td>
      <td>85.74%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>93035</td>
      <td>93039</td>
      <td>+4</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15469</td>
      <td>15469</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26072/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26072/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>26072a</th>
      <th colspan=4>#26072 72293b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.87%</td>
      <td>84.86%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12572</td>
      <td>12565</td>
      <td>-7</td>
      <td>87</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2241</td>
      <td>2241</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26072/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26072/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26072/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on modules/navier_stokes/include/materials/INSADTauMaterial.h:
```diff
@@ -308,28 +308,34 @@ INSADTauMaterialTempl<T>::viscousTermRZ()
+    rz_term(_rz_radial_coord) += _mu[_qp] * _velocity[_qp](_rz_radial_coord) / (r * r);
+    _viscous_strong_residual[_qp] += rz_term;
+  }
+  if (_viscous_form == NS::ViscousForm::Traction)
```
do we have docs about this traction form?
if so it d be nice to link to it from all the 'Traction' objects

GiudGiud commented on modules/navier_stokes/src/kernels/INSADMomentumViscous.C:
```diff
@@ -57,15 +57,16 @@ INSADMomentumViscous::qpViscousTerm()
-    resid *= 2.;
+    extra_term *= 2.;
 
-  if (_use_displaced_mesh)
```
this was just an optimization right

lindsayad replied:
Some unknown degree of optimization yes. My hope is that multiplying dual numbers without derivatives is not too expensive in the scheme of things

GiudGiud commented on modules/navier_stokes/src/kernels/INSBase.C:
```diff
@@ -326,3 +328,69 @@ INSBase::dTauDUComp(unsigned comp)
+  rz_term(1) = -_mu[_qp] * _grad_v_vel[_qp](_rz_radial_coord) / r;
+  mooseAssert((_rz_radial_coord == 0) || (_rz_radial_coord == 1),
+              "We expect X or Y as the possible radial coordinate");
+  if (_rz_radial_coord == 0)
+    rz_term(0) += _mu[_qp] * _u_vel[_qp] / (r * r);
+  else
+    rz_term(1) += _mu[_qp] * _v_vel[_qp] / (r * r);
```
can one line this?
with 
rz_term(_rz_radial_coord)

lindsayad replied:
It uses different variables

GiudGiud replied:
ah good point.

GiudGiud commented on modules/navier_stokes/src/kernels/INSMomentumTractionFormRZ.C:
```diff
@@ -98,9 +86,9 @@ INSMomentumTractionFormRZ::computeQpOffDiagJacobian(unsigned jvar)
+  if (jvar == _p_var_number && _component == _rz_radial_coord && _integrate_p_by_parts)
   {
-    const Real r = _q_point[_qp](0);
+    const auto r = _q_point[_qp](_rz_radial_coord);
```
want to use ad_q_point for displaced mesh Jacobian or not?

lindsayad replied:
Not really for the hand-coded objects. I've made no effort to support 100% accurate Jacobians for displaced meshes in the hand-coded objects to date. If wanting to add that support, I think it should happen in a larger sweep.

I'm still debating whether just to remove the INS objects in general. However, they have been very helpful in checking for bugs in the INSAD implementation. In an ideal world I would use the INSFE objects for that ... or remove the INSFE objects.

lindsayad replied:
Also in that same larger discussion, for non-displaced meshes the INS Jacobians are 100% accurate and so the INS objects have faster run-times than INSAD so that would be a good argument for keeping them around. Anyway this is not on-topic haha

GiudGiud replied:
ah ok that makes sense. 
INSAD will use ad_q_point and INS should consistently use _q_point and manually do the derivative later if it s ever needed

yeah INSFE vs INS will have to be decided some day

GiudGiud reviewed (commented):
will look more later

lindsayad committed:
Remove note about inconsistent stabilization with INSAD

We now offer a fully consistent form due to using second order
scalar field Lagrange FEs behind-the-scenes

lindsayad committed:
Add more documentation on laplace/traction viscous forms

lindsayad committed:
MooseDocs doesn't seem to like lable reference from aligned env

lindsayad commented on modules/navier_stokes/src/kernels/INSBase.C:
```diff
@@ -326,3 +328,69 @@ INSBase::dTauDUComp(unsigned comp)
+  const auto & grad_r_vel = (_rz_radial_coord == 0) ? _grad_u_vel[_qp] : _grad_v_vel[_qp];
+  const auto & r_vel = (_rz_radial_coord == 0) ? _u_vel[_qp] : _v_vel[_qp];
+  ret += -_mu[_qp] * grad_r_vel / r;
+  ret(_rz_radial_coord) += _mu[_qp] * r_vel / (r * r);
```
what you can be confident about now is that if something is wrong, then it's wrong in two places! 😄

GiudGiud commented on modules/navier_stokes/src/kernels/INSBase.C:
```diff
@@ -143,7 +145,7 @@ INSBase::strongViscousTermLaplace()
 RealVectorValue
```
in the routine above, are we missing a grad_u * grad_mu term here?

from div(mu grad_u) = mu div grad u + grad_u . grad_mu

GiudGiud replied:
I guess this isnt specific to RZ. Porbably a common approximation considering mu doesnt move much. Could be worth documenting?

GiudGiud replied:
yes that is also true that we have not necessarily paid more attention to this term elsewhere

lindsayad replied:
I've opened https://github.com/idaholab/moose/issues/26217

GiudGiud reviewed (approved):
all good
